text/0000-stabilize-marker-freeze.md

```diff
@@ -0,0 +1,149 @@
+
+- Some people have previously argued that this would be akin to exposing compiler internals.
+	- The RFC author disagrees, viewing `Freeze` in a similar light as `Send` and `Sync`: a trait that allows soundness requirements to be proven at compile time.
+- `Freeze` being an auto-trait, it is, like `Send` and `Sync` a sneaky SemVer hazard.
+	- Note that this SemVer hazard already exists through the existence of static-promotion, as exemplified by the following example:
+	```rust
+    // old version of the crate.
+    mod v1 {
+        pub struct S(i32);
+        impl S {
+            pub const fn new() -> Self { S(42) }
+        }
+    }
+
+    // new version of the crate, adding interior mutability.
+    mod v2 {
+        use std::cell::Cell;
+        pub struct S(Cell<i32>);
+        impl S {
+            pub const fn new() -> Self { S(Cell::new(42)) }
+        }
+    }
+
+    // Old version: builds
+    const C1: &v1::S = &v1::S::new();
+    // New version: does not build
+    const C2: &v2::S = &v2::S::new();
+	```
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- The benefits of stabilizing `core::mem::Freeze` have been highlighted in [Motivation](#motivation).
+	- By not stabilizing `core::mem::Freeze` in trait bounds, we are preventing useful and sound code patterns from existing which were previously supported.
+- Alternatively, a non-auto sub-trait of `core::mem::Freeze` may be defined:
+	- While this reduces the SemVer hazard by making its breakage more obvious, this does lose part of the usefulness that `core::mem::Freeze` would provide to projects such as `zerocopy`.
+	- A "perfect" derive macro should then be introduced to ease the implementation of this trait. A lint may be introduced in `clippy` to inform users of the existence and applicability of this new trait.
+
+# Prior art
+[prior-art]: #prior-art
+
+- This feature has been available in `nightly` for 7 years, and is used internally by the compiler.
+- The work necessary for this RFC has already been done and merged in [this PR](https://github.com/rust-lang/rust/issues/121675), and a [tracking issue](https://github.com/rust-lang/rust/issues/121675) was opened.
+- zerocopy's [`Immutable`](https://docs.rs/zerocopy/0.8.0-alpha.11/zerocopy/trait.Immutable.html) seeks to provide the same guarantees as `core::marker::Freeze`.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+[Should the trait be exposed under a different name?](https://github.com/rust-lang/rust/pull/121501#issuecomment-1962900148)
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- One might later consider whether `core::mem::Freeze` should be allowed to be `unsafe impl`'d like `Send` and `Sync` are, possibly allowing wrappers around interiorly mutable data to hide this interior mutability from constructs that require `Freeze` if the logic surrounding it guarantees that the interior mutability will never be used.
```

Comment:
If extern types ever get stabilised they would present a situation where `unsafe impl Freeze` would be useful. Extern types represent types opaque to the Rust compiler (usually because they're defined in another language and accessed via FFI) and so they have to be `!Freeze` by default. I'm not sure that's enough motivation to change the approach for this RFC, but it does present a much simpler and less scary reason to want to implement Freeze for a type

Reply:
An `extern type T` can never appear directly in a struct layout, you must use indirection like `*const T` and `&T`, but a reference/pointer itself is `Freeze`, so whether `T: Freeze` or not does not really matter.

Regarding the motivation (to properly address rust-lang/rust#121250, well you can't dereference a `T`-pointer to create a value `T` in Rust, so the motivation doesn't apply to extern type at all.

Reply:
Ahhh that's a very good point, don't mind me just gonna go steal that and write it in the extern types RFC. 😁

Reply:
> An `extern type T` can never appear directly in a struct layout, you must use indirection like `*const T` and `&T`, but a reference/pointer itself is `Freeze`, so whether `T: Freeze` or not does not really matter.

I hope Rust will support extern types in structs in the future, e.g.:

```rust
extern {
    pub type Body;
}

unsafe impl Aligned for Body {
    const ALIGN: usize = 8;
}

#[repr(C)]
pub struct MyStruct {
    tag: u8,
    body: Body, // fine since rust knows what alignment is needed
}
```

Reply:
@programmerjake In what scenario such a `MyStruct` will appear? AFAIK you can't use an incomplete type in C or C++.

```c
#include <stdint.h>

typedef struct Body Body;

typedef struct MyStruct {
    uint8_t tag;
    _Alignas(8) Body body;
    //^~ error: field 'body' has incomplete type
} MyStruct;
```

Reply:
> @programmerjake In what scenario such a `MyStruct` will appear? AFAIK you can't use an incomplete type in C or C++.

In C it would be written something like:
```c
#include <stdint.h>

typedef struct MyStruct {
    uint8_t tag;
    _Alignas(8) char body[]; // always cast to pointer to whatever actual body type is used before use
} MyStruct;
```
or
```c
typedef struct MyStruct {
    uint8_t tag;
    char padding[7];
    // _Alignas(8) T body;
} MyStruct;
```

Reply:
an example of how it's done in C is Vulkan's [`VkBaseInStructure`](https://github.com/KhronosGroup/Vulkan-Headers/blob/5677bafb820e476441e9e1f745371b72133407d3/include/vulkan/vulkan_core.h#L2856-L2859) struct, which is a prefix of basically every struct used in Vulkan's API.

Reply:
This is the Freeze RFC, not the extern type RFC. :)

I don't think this RFC has to take into account extensions to unstable features.

Reply:
@programmerjake if you need a `char[]` on the C side you should also use a "thin slice" on the Rust side.

While `VkBaseInStructure` is indeed the "base class" of many other Vk classes, I don't think it is appropriate to insert the "tail" into `VkBaseInStructure` itself. (Typical proposal for Rust to support OOP are like #2431) Besides, even if you have that "extern type tail" you can't realize a `VkBaseInStructure` as a concrete subclass type like `VkMemoryBarrier` anyway.

Reply:
I'd argue this purpose would be much better satisfied using a trait (the sacrilegious maniac in me wants to say `Deref` would even do the trick) that lets you access that "base class" (either by casting the reference or by calling a library-defined function, such that the offset may be changed by the library if suitable), rather than introduce a language feature that wouldn't really bring much to the table.

Regardless, this has little to do with this RFC, as the status of `unsafe impl Freeze` is _deliberately_ left out of scope. I'll resolve this conversation to reduce noise :)